src/modules/rlm_kafka/rlm_kafka.c

```diff
@@ -377,29 +512,56 @@ static ssize_t kafka_xlat(void *instance, REQUEST *request, char const *fmt, cha
 	entry = rbtree_finddata(inst->rkt_by_name_tree, &my_topic);
 	if (!entry || !entry->rkt) {
 		RWARN("No configuration section exists for kafka topic \"%s\"", buf);
-		return -1;
+		goto error;
+	}
+
+	/*
+	 *  Extract the header specification, and generate the headers
+	 *
+	 */
+	q = p;
+	if (*p == '(') {
+		p = strchr(p, ')');
+		if (!p) {
+			REDEBUG("Header list is missing closing parenthesis)");
+			goto error;
+		}
+		MEM(headers = talloc_strndup(NULL, q + 1, p - q - 1));
+		if (*headers && create_headers(request, headers, &hdrs) < 0) {
+			RDEBUG3("Failed to create headers");
+			talloc_free(headers);
+			goto error;
+		}
+		talloc_free(headers);
+		p++;
+		if (*p != ' ') {
+			REDEBUG("Kafka xlat must begin with a topic, optionally followed by headers, then the payload");
+			goto error;
+		}
+		p++;
 	}
 
 	/*
 	 *  Extract the key, if there is one, otherwise expect a space.
 	 *
```

Comment:
I wonder if it would be nicer to always expect the key and to be in e.g. `[]`, rather than requiring a practically invisible double space?

Reply:
Agreed to retain current API and potentially rework things for v4 based on experience and any new features.